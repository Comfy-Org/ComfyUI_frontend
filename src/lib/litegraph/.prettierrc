{
  "singleQuote": false,
  "semi": false,
  "tabWidth": 2,
  "trailingComma": "all",
  "overrides": [{ "files": "*.ts", "options": { "requirePragma": true } }],
  "arrowParens": "avoid"
}
