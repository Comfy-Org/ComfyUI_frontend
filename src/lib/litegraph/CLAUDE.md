- This codebase has extensive eslint autofix rules and IDEs are configured to use eslint as the format on save tool. Run ESLint instead of manually figuring out whitespace fixes or other trivial style concerns. Review the results and correct any remaining eslint errors.
- Take advantage of `TypedArray` `subarray` when appropriate.
- The `size` and `pos` properties of `Rectangle` share the same array buffer (`subarray`); they may be used to set the rectangles size and position.
- Prefer single line `if` syntax over adding curly braces, when the statement has a very concise expression and concise, single line statement.
- Do not replace `&&=` or `||=` with `=` when there is no reason to do so. If you do find a reason to remove either `&&=` or `||=`, leave a comment explaining why the removal occurred.
- You are allowed to research code on https://developer.mozilla.org/ and https://stackoverflow.com without asking.
- When adding features, always write vitest unit tests using cursor rules in @.cursor
- When writing methods, prefer returning idiomatic JavaScript `undefined` over `null`.

# Bash commands

- `npm run typecheck` Run the typechecker
- `npm run build` Build the project
- `npm run lint:fix` Run ESLint

# Code style

- Always prefer best practices when writing code.
- Write using concise, legible, and easily maintainable code.
- Avoid repetition where possible, but not at the expense of code legibility.
- Type assertions are an absolute last resort. In almost all cases, they are a crutch that leads to brittle code.

# Workflow

- Be sure to typecheck when youâ€™re done making a series of code changes
- Prefer running single tests, and not the whole test suite, for performance
