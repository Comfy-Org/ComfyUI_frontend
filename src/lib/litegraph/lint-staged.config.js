export default {
  "*.css": stagedFiles => `prettier --write ${stagedFiles.join(" ")}`,

  "*.js": stagedFiles => prettierAndEslint(stagedFiles),

  "*.{ts,mts}": stagedFiles => [
    ...prettierAndEslint(stagedFiles),
    "tsc",
  ],
}

function prettierAndEslint(fileNames) {
  return [
    `prettier --write ${fileNames.join(" ")}`,
    `eslint --fix ${fileNames.join(" ")}`,
  ]
}
