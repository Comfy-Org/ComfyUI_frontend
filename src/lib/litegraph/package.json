{
  "name": "@comfyorg/litegraph",
  "version": "0.16.3",
  "type": "module",
  "description": "A graph node editor similar to PD or UDK Blueprints. It works in an HTML5 Canvas and allows to export graphs to be included in applications.",
  "main": "./dist/litegraph.umd.js",
  "module": "./dist/litegraph.es.js",
  "types": "./dist/litegraph.d.ts",
  "style": "./dist/css/litegraph.css",
  "exports": {
    ".": {
      "types": "./dist/litegraph.d.ts",
      "import": "./dist/litegraph.es.js",
      "require": "./dist/litegraph.umd.js"
    },
    "./style.css": "./dist/css/litegraph.css",
    "./dist/*": "./dist/*.d.ts"
  },
  "directories": {
    "doc": "doc"
  },
  "private": false,
  "scripts": {
    "build": "tsc && vite build",
    "dev": "vite",
    "preview": "vite preview",
    "watch": "vite build --watch",
    "test": "vitest",
    "lint": "eslint",
    "lint:fix": "eslint --fix",
    "lint:ci": "eslint src",
    "format": "prettier --check './src/*.{js,ts,tsx,vue,mts}'",
    "format:fix": "prettier --write './src/*.{js,ts,tsx,vue,mts}'",
    "typecheck": "tsc"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:Comfy-Org/litegraph.js.git"
  },
  "author": "comfyorg",
  "license": "MIT",
  "files": [
    "dist"
  ],
  "bugs": {
    "url": "https://github.com/Comfy-Org/litegraph.js/issues"
  },
  "homepage": "https://github.com/Comfy-Org/litegraph.js",
  "devDependencies": {
    "@types/node": "^22.13.9",
    "jsdom": "^25.0.1",
    "ts-node": "^10.9.2",
    "vite-plugin-dts": "^4.5.4",
  }
}
