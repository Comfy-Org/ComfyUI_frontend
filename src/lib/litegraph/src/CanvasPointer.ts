import type { CompassCorners } from './interfaces'
import { dist2 } from './measure'
import type { CanvasPointerEvent } from './types/events'

/**
 * Allows click and drag actions to be declared ahead of time during a pointerdown event.
 *
 * By default, it retains the most recent event of each type until it is reset (on pointerup).
 * - {@link eDown}
 * - {@link eMove}
 * - {@link eUp}
 *
 * Depending on whether the user clicks or drags the pointer, only the appropriate callbacks are called:
 * - {@link onClick}
 * - {@link onDoubleClick}
 * - {@link onDragStart}
 * - {@link onDrag}
 * - {@link onDragEnd}
 * - {@link finally}
 * @see
 * - {@link LGraphCanvas.processMouseDown}
 * - {@link LGraphCanvas.processMouseMove}
 * - {@link LGraphCanvas.processMouseUp}
 */
export class CanvasPointer {
  /** Maximum time in milliseconds to ignore click drift */
  static bufferTime = 150

  /** Maximum gap between pointerup and pointerdown events to be considered as a double click */
  static doubleClickTime = 300

  /** Maximum offset from click location */
  static get maxClickDrift() {
    return this.#maxClickDrift
  }

  static set maxClickDrift(value) {
    this.#maxClickDrift = value
    this.#maxClickDrift2 = value * value
  }

  static #maxClickDrift = 6
  /** {@link maxClickDrift} squared.  Used to calculate click drift without `sqrt`. */
  static #maxClickDrift2 = this.#maxClickDrift ** 2

  /** Assume that "wheel" events with both deltaX and deltaY less than this value are trackpad gestures. */
  static trackpadThreshold = 60

  /**
   * The minimum time between "wheel" events to allow switching between trackpad
   * and mouse modes.
   *
   * This prevents trackpad "flick" panning from registering as regular mouse wheel.
   * After a flick gesture is complete, the automatic wheel events are sent with
   * reduced frequency, but much higher deltaX and deltaY values.
   */
  static trackpadMaxGap = 200

  /** The element this PointerState should capture input against when dragging. */
  element: Element
  /** Pointer ID used by drag capture. */
  pointerId?: number

  /** Set to true when if the pointer moves far enough after a down event, before the corresponding up event is fired. */
  dragStarted: boolean = false

  /** The {@link eUp} from the last successful click */
  eLastDown?: CanvasPointerEvent

  /** Used downstream for touch event support. */
  isDouble: boolean = false
  /** Used downstream for touch event support. */
  isDown: boolean = false

  /** The resize handle currently being hovered or dragged */
  resizeDirection?: CompassCorners

  /**
   * If `true`, {@link eDown}, {@link eMove}, and {@link eUp} will be set to
   * `undefined` when {@link reset} is called.
   *
   * Default: `true`
   */
  clearEventsOnReset: boolean = true

  /** The last pointerdown event for the primary button */
  eDown?: CanvasPointerEvent
  /** The last pointermove event for the primary button */
  eMove?: CanvasPointerEvent
  /** The last pointerup event for the primary button */
  eUp?: CanvasPointerEvent

  /** The last pointermove event that was treated as a trackpad gesture. */
  lastTrackpadEvent?: WheelEvent

  /**
   * If set, as soon as the mouse moves outside the click drift threshold, this action is run once.
   * @param pointer [DEPRECATED] This parameter will be removed in a future release.
   * @param eMove The pointermove event of this ongoing drag action.
   *
   * It is possible for no `pointermove` events to occur, but still be far from
   * the original `pointerdown` event. In this case, {@link eMove} will be null, and
   * {@link onDragEnd} will be called immediately after {@link onDragStart}.
   */
  onDragStart?(pointer: this, eMove?: CanvasPointerEvent): unknown

  /**
   * Called on pointermove whilst dragging.
   * @param eMove The pointermove event of this ongoing drag action
   */
  onDrag?(eMove: CanvasPointerEvent): unknown

  /**
   * Called on pointerup after dragging (i.e. not called if clicked).
   * @param upEvent The pointerup or pointermove event that triggered this callback
   */
  onDragEnd?(upEvent: CanvasPointerEvent): unknown

  /**
   * Callback that will be run once, the next time a pointerup event appears to be a normal click.
   * @param upEvent The pointerup or pointermove event that triggered this callback
   */
  onClick?(upEvent: CanvasPointerEvent): unknown

  /**
   * Callback that will be run once, the next time a pointerup event appears to be a normal click.
   * @param upEvent The pointerup or pointermove event that triggered this callback
   */
  onDoubleClick?(upEvent: CanvasPointerEvent): unknown

  /**
   * Run-once callback, called at the end of any click or drag, whether or not it was successful in any way.
   *
   * The setter of this callback will call the existing value before replacing it.
   * Therefore, simply setting this value twice will execute the first callback.
   */
  get finally() {
    return this.#finally
  }

  set finally(value) {
    try {
      this.#finally?.()
    } finally {
      this.#finally = value
    }
  }

  #finally?: () => unknown

  constructor(element: Element) {
    this.element = element
  }

  /**
   * Callback for `pointerdown` events.  To be used as the event handler (or called by it).
   * @param e The `pointerdown` event
   */
  down(e: CanvasPointerEvent): void {
    this.reset()
    this.eDown = e
    this.pointerId = e.pointerId
    this.element.setPointerCapture(e.pointerId)
  }

  /**
   * Callback for `pointermove` events.  To be used as the event handler (or called by it).
   * @param e The `pointermove` event
   */
  move(e: CanvasPointerEvent): void {
    const { eDown } = this
    if (!eDown) return

    // No buttons down, but eDown exists - clean up & leave
    if (!e.buttons) {
      this.reset()
      return
    }

    // Primary button released - treat as pointerup.
    if (!(e.buttons & eDown.buttons)) {
      this.#completeClick(e)
      this.reset()
      return
    }
    this.eMove = e
    this.onDrag?.(e)

    // Dragging, but no callback to run
    if (this.dragStarted) return

    const longerThanBufferTime =
      e.timeStamp - eDown.timeStamp > CanvasPointer.bufferTime
    if (longerThanBufferTime || !this.#hasSamePosition(e, eDown)) {
      this.#setDragStarted(e)
    }
  }

  /**
   * Callback for `pointerup` events.  To be used as the event handler (or called by it).
   * @param e The `pointerup` event
   */
  up(e: CanvasPointerEvent): boolean {
    if (e.button !== this.eDown?.button) return false

    this.#completeClick(e)
    const { dragStarted } = this
    this.reset()
    return !dragStarted
  }

  #completeClick(e: CanvasPointerEvent): void {
    const { eDown } = this
    if (!eDown) return

    this.eUp = e

    if (this.dragStarted) {
      // A move event already started drag
      this.onDragEnd?.(e)
    } else if (!this.#hasSamePosition(e, eDown)) {
      // Teleport without a move event (e.g. tab out, move, tab back)
      this.#setDragStarted()
      this.onDragEnd?.(e)
    } else if (this.onDoubleClick && this.#isDoubleClick()) {
      // Double-click event
      this.onDoubleClick(e)
      this.eLastDown = undefined
    } else {
      // Normal click event
      this.onClick?.(e)
      this.eLastDown = eDown
    }
  }

  /**
   * Checks if two events occurred near each other - not further apart than the maximum click drift.
   * @param a The first event to compare
   * @param b The second event to compare
   * @param tolerance2 The maximum distance (squared) before the positions are considered different
   * @returns `true` if the two events were no more than {@link maxClickDrift} apart, otherwise `false`
   */
  #hasSamePosition(
    a: PointerEvent,
    b: PointerEvent,
    tolerance2 = CanvasPointer.#maxClickDrift2
  ): boolean {
    const drift = dist2(a.clientX, a.clientY, b.clientX, b.clientY)
    return drift <= tolerance2
  }

  /**
   * Checks whether the pointer is currently past the max click drift threshold.
   * @returns `true` if the latest pointer event is past the the click drift threshold
   */
  #isDoubleClick(): boolean {
    const { eDown, eLastDown } = this
    if (!eDown || !eLastDown) return false

    // Use thrice the drift distance for double-click gap
    const tolerance2 = (3 * CanvasPointer.#maxClickDrift) ** 2
    const diff = eDown.timeStamp - eLastDown.timeStamp
    return (
      diff > 0 &&
      diff < CanvasPointer.doubleClickTime &&
      this.#hasSamePosition(eDown, eLastDown, tolerance2)
    )
  }

  #setDragStarted(eMove?: CanvasPointerEvent): void {
    this.dragStarted = true
    this.onDragStart?.(this, eMove)
    delete this.onDragStart
  }

  /**
   * Checks if the given wheel event is part of a continued trackpad gesture.
   * @param e The wheel event to check
   * @returns `true` if the event is part of a continued trackpad gesture, otherwise `false`
   */
  #isContinuationOfGesture(e: WheelEvent): boolean {
    const { lastTrackpadEvent } = this
    if (!lastTrackpadEvent) return false

    return (
      e.timeStamp - lastTrackpadEvent.timeStamp < CanvasPointer.trackpadMaxGap
    )
  }

  /**
   * Checks if the given wheel event is part of a trackpad gesture.
   * @param e The wheel event to check
   * @returns `true` if the event is part of a trackpad gesture, otherwise `false`
   */
  isTrackpadGesture(e: WheelEvent): boolean {
    if (this.#isContinuationOfGesture(e)) {
      this.lastTrackpadEvent = e
      return true
    }

    const threshold = CanvasPointer.trackpadThreshold
    return Math.abs(e.deltaX) < threshold && Math.abs(e.deltaY) < threshold
  }

  /**
   * Resets the state of this {@link CanvasPointer} instance.
   *
   * The {@link finally} callback is first executed, then all callbacks and intra-click
   * state is cleared.
   */
  reset(): void {
    // The setter executes the callback before clearing it
    this.finally = undefined
    delete this.onClick
    delete this.onDoubleClick
    delete this.onDragStart
    delete this.onDrag
    delete this.onDragEnd

    this.isDown = false
    this.isDouble = false
    this.dragStarted = false
    this.resizeDirection = undefined

    if (this.clearEventsOnReset) {
      this.eDown = undefined
      this.eMove = undefined
      this.eUp = undefined
    }

    const { element, pointerId } = this
    this.pointerId = undefined
    if (typeof pointerId === 'number' && element.hasPointerCapture(pointerId)) {
      element.releasePointerCapture(pointerId)
    }
  }
}
