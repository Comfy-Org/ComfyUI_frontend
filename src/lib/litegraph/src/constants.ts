/**
 * Subgraph constants
 *
 * This entire module is exported as `Constants`.
 */

/** ID of the virtual input node of a subgraph. */
export const SUBGRAPH_INPUT_ID = -10

/** ID of the virtual output node of a subgraph. */
export const SUBGRAPH_OUTPUT_ID = -20
