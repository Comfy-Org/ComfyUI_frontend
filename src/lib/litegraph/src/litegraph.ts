import type { ContextMenu } from "./ContextMenu"
import type { ConnectingLink, Point } from "./interfaces"
import type {
  IContextMenuOptions,
  INodeSlot,
  Size,
} from "./interfaces"
import type { LGraphNode } from "./LGraphNode"
import type { CanvasEventDetail } from "./types/events"
import type { RenderShape, TitleMode } from "./types/globalEnums"

// Must remain above LiteGraphGlobal (circular dependency due to abstract factory behaviour in `configure`)
export { Subgraph } from "./subgraph/Subgraph"

import { LiteGraphGlobal } from "./LiteGraphGlobal"
import { loadPolyfills } from "./polyfills"

export const LiteGraph = new LiteGraphGlobal()

// Load legacy polyfills
loadPolyfills()

// Backwards compat

// Type definitions for litegraph.js 0.7.0
// Project: litegraph.js
// Definitions by: NateScarlet <https://github.com/NateScarlet>
/** @deprecated Use {@link Point} instead. */
export type Vector2 = Point
/** @deprecated Use {@link Rect} instead. */
export type Vector4 = [number, number, number, number]

export interface IContextMenuItem {
  content: string
  callback?: ContextMenuEventListener
  /** Used as innerHTML for extra child element */
  title?: string
  disabled?: boolean
  has_submenu?: boolean
  submenu?: {
    options: IContextMenuItem[]
  } & IContextMenuOptions
  className?: string
}

export type ContextMenuEventListener = (
  value: IContextMenuItem,
  options: IContextMenuOptions,
  event: MouseEvent,
  parentMenu: ContextMenu<unknown> | undefined,
  node: LGraphNode,
) => boolean | void

export interface LinkReleaseContext {
  node_to?: LGraphNode
  node_from?: LGraphNode
  slot_from: INodeSlot
  type_filter_in?: string
  type_filter_out?: string
}

export interface LinkReleaseContextExtended {
  links: ConnectingLink[]
}

export interface LiteGraphCanvasEvent extends CustomEvent<CanvasEventDetail> {}

export interface LGraphNodeConstructor<T extends LGraphNode = LGraphNode> {
  new (title: string, type?: string): T

  title: string
  type: string
  size?: Size
  min_height?: number
  slot_start_y?: number
  widgets_info?: any
  collapsable?: boolean
  color?: string
  bgcolor?: string
  shape?: RenderShape
  title_mode?: TitleMode
  title_color?: string
  title_text_color?: string
  keepAllLinksOnBypass: boolean
  resizeHandleSize?: number
  resizeEdgeSize?: number
}

// End backwards compat

export { InputIndicators } from "./canvas/InputIndicators"
export { LinkConnector } from "./canvas/LinkConnector"
export { isOverNodeInput, isOverNodeOutput } from "./canvas/measureSlots"
export { CanvasPointer } from "./CanvasPointer"
export * as Constants from "./constants"
export { ContextMenu } from "./ContextMenu"
export { CurveEditor } from "./CurveEditor"
export { DragAndScale } from "./DragAndScale"
export { LabelPosition, SlotDirection, SlotShape, SlotType } from "./draw"
export { strokeShape } from "./draw"
export { Rectangle } from "./infrastructure/Rectangle"
export type {
  CanvasColour,
  ColorOption,
  ConnectingLink,
  Direction,
  IBoundaryNodes,
  IColorable,
  IContextMenuOptions,
  IContextMenuValue,
  IFoundSlot,
  IInputOrOutput,
  INodeFlags,
  INodeInputSlot,
  INodeOutputSlot,
  INodeSlot,
  ISlotType,
  KeysOfType,
  LinkNetwork,
  LinkSegment,
  MethodNames,
  PickByType,
  Point,
  Positionable,
  ReadonlyLinkNetwork,
  ReadOnlyPoint,
  ReadOnlyRect,
  Rect,
  Size,
} from "./interfaces"
export { LGraph } from "./LGraph"
export { BadgePosition, LGraphBadge, type LGraphBadgeOptions } from "./LGraphBadge"
export { LGraphCanvas, type LGraphCanvasState } from "./LGraphCanvas"
export { LGraphGroup } from "./LGraphGroup"
export { LGraphNode, type NodeId } from "./LGraphNode"
export { COMFY_VUE_NODE_DIMENSIONS, type ComponentHeightKey } from "./LiteGraphGlobal"
export { LGraphNodeProperties } from "./LGraphNodeProperties"
export { type LinkId, LLink } from "./LLink"
export { clamp, createBounds } from "./measure"
export { Reroute, type RerouteId } from "./Reroute"
export { type ExecutableLGraphNode, ExecutableNodeDTO } from "./subgraph/ExecutableNodeDTO"
export { SubgraphNode } from "./subgraph/SubgraphNode"
export type { CanvasPointerEvent } from "./types/events"
export {
  CanvasItem,
  EaseFunction,
  LGraphEventMode,
  LinkMarkerShape,
  RenderShape,
  TitleMode,
} from "./types/globalEnums"
export type {
  ExportedSubgraph,
  ExportedSubgraphInstance,
  ExportedSubgraphIONode,
  ISerialisedGraph,
  SerialisableGraph,
  SerialisableLLink,
  SubgraphIO,
} from "./types/serialisation"
export type { IWidget } from "./types/widgets"
export { isColorable } from "./utils/type"
export { createUuidv4 } from "./utils/uuid"
export { BaseSteppedWidget } from "./widgets/BaseSteppedWidget"
export { BaseWidget } from "./widgets/BaseWidget"
export { BooleanWidget } from "./widgets/BooleanWidget"
export { ButtonWidget } from "./widgets/ButtonWidget"
export { ComboWidget } from "./widgets/ComboWidget"
export { KnobWidget } from "./widgets/KnobWidget"
export { LegacyWidget } from "./widgets/LegacyWidget"
export { NumberWidget } from "./widgets/NumberWidget"
export { SliderWidget } from "./widgets/SliderWidget"
export { TextWidget } from "./widgets/TextWidget"
export { isComboWidget } from "./widgets/widgetMap"
