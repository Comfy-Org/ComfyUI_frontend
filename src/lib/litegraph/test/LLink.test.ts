import { describe, expect, it, vi } from 'vitest'

import { LGraph, LGraphNode, LLink } from '@/lib/litegraph/src/litegraph'

import { test } from './testExtensions'

describe('LLink', () => {
  test('matches previous snapshot', () => {
    const link = new LLink(1, 'float', 4, 2, 5, 3)
    expect(link.serialize()).toMatchSnapshot('Basic')
  })

  test('serializes to the previous snapshot', () => {
    const link = new LLink(1, 'float', 4, 2, 5, 3)
    expect(link.serialize()).toMatchSnapshot('Basic')
  })

  describe('disconnect', () => {
    it('should clear the target input link reference when disconnecting', () => {
      // Create a graph and nodes
      const graph = new LGraph()
      const sourceNode = new LGraphNode('Source')
      const targetNode = new LGraphNode('Target')

      // Add nodes to graph
      graph.add(sourceNode)
      graph.add(targetNode)

      // Add slots
      sourceNode.addOutput('out', 'number')
      targetNode.addInput('in', 'number')

      // Connect the nodes
      const link = sourceNode.connect(0, targetNode, 0)
      expect(link).toBeDefined()
      expect(targetNode.inputs[0].link).toBe(link?.id)

      // Mock setDirtyCanvas
      const setDirtyCanvasSpy = vi.spyOn(targetNode, 'setDirtyCanvas')

      // Disconnect the link
      link?.disconnect(graph)

      // Verify the target input's link reference is cleared
      expect(targetNode.inputs[0].link).toBeNull()

      // Verify setDirtyCanvas was called
      expect(setDirtyCanvasSpy).toHaveBeenCalledWith(true, false)
    })

    it('should handle disconnecting when target node is not found', () => {
      // Create a link with invalid target
      const graph = new LGraph()
      const link = new LLink(1, 'number', 1, 0, 999, 0) // Invalid target id

      // Should not throw when disconnecting
      expect(() => link.disconnect(graph)).not.toThrow()
    })

    it('should only clear link reference if it matches the current link id', () => {
      // Create a graph and nodes
      const graph = new LGraph()
      const sourceNode1 = new LGraphNode('Source1')
      const sourceNode2 = new LGraphNode('Source2')
      const targetNode = new LGraphNode('Target')

      // Add nodes to graph
      graph.add(sourceNode1)
      graph.add(sourceNode2)
      graph.add(targetNode)

      // Add slots
      sourceNode1.addOutput('out', 'number')
      sourceNode2.addOutput('out', 'number')
      targetNode.addInput('in', 'number')

      // Create first connection
      const link1 = sourceNode1.connect(0, targetNode, 0)
      expect(link1).toBeDefined()

      // Disconnect first connection
      targetNode.disconnectInput(0)

      // Create second connection
      const link2 = sourceNode2.connect(0, targetNode, 0)
      expect(link2).toBeDefined()
      expect(targetNode.inputs[0].link).toBe(link2?.id)

      // Try to disconnect the first link (which is already disconnected)
      // It should not affect the current connection
      link1?.disconnect(graph)

      // The input should still have the second link
      expect(targetNode.inputs[0].link).toBe(link2?.id)
    })
  })
})
