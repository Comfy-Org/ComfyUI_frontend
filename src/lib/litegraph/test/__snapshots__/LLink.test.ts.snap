// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LLink > matches previous snapshot > Basic 1`] = `
[
  1,
  4,
  2,
  5,
  3,
  "float",
]
`;

exports[`LLink > serializes to the previous snapshot > Basic 1`] = `
[
  1,
  4,
  2,
  5,
  3,
  "float",
]
`;
